csrc/dynamic_type.h

```diff
@@ -403,15 +403,135 @@ DEFINE_BINARY_OP(land, &&);
 DEFINE_BINARY_OP(lor, ||);
 DEFINE_BINARY_OP(lshift, <<);
 DEFINE_BINARY_OP(rshift, >>);
-DEFINE_BINARY_OP(eq, ==);
-DEFINE_BINARY_OP(neq, !=);
-DEFINE_BINARY_OP(lt, <);
-DEFINE_BINARY_OP(gt, >);
-DEFINE_BINARY_OP(le, <=);
-DEFINE_BINARY_OP(ge, >=);
 
 #undef DEFINE_BINARY_OP
 
+#define DEFINE_COMPARE_OP(opname, op)                                         \
+  /*TODO: we should inline the definition of lambdas into enable_if,*/        \
+  /*but I can only do this in C++20 */                                        \
+  constexpr auto opname##_defined_checker = [](auto x, auto y) constexpr {    \
+    using X = typename decltype(x)::type;                                     \
+    using Y = typename decltype(y)::type;                                     \
+    if constexpr (opcheck<X> op opcheck<Y>) {                                 \
+      return std::is_convertible_v<                                           \
```

Comment:
Note that `std::is_convertible_v<From, To>`, is different from `opcheck<From>.canCastTo(opcheck<To>)`. The former is true only when `From` can be **implicitly** converted to `To`, while the latter is true also for explicit casts. For this case here, only allowing implicit casting makes more sense.